quic/api.nim

```diff
@@ -21,36 +22,69 @@ export close
 export waitClosed
 export errors
 export destroy
+export CertificateVerifier
+export certificateVerifierCB
+export CustomCertificateVerifier
+export InsecureCertificateVerifier
+export init
 
 type TLSConfig* = object
-  certificate*: seq[byte]
-  key*: seq[byte]
-  # verifyCertificate = Opt[proc] # None to skip verif
-  # 
+  certificate: seq[byte]
+  key: seq[byte]
+  certificateVerifier: Opt[CertificateVerifier]
 
-proc init*(t: typedesc[TLSConfig], certificate, key: seq[byte]): TLSConfig =
-  return TLSConfig(certificate: certificate, key: key)
+type Quic = ref object of RootObj
+  tlsConfig: TLSConfig
 
-proc listen*(
-    address: TransportAddress, tlsConfig: TLSConfig
-): Listener {.raises: [QuicConfigError, QuicError, TransportOsError].} =
+type QuicClient* = object of Quic
+
+type QuicServer* = object of Quic
+
+proc init*(
+    t: typedesc[TLSConfig],
+    certificate: seq[byte] = @[],
+    key: seq[byte] = @[],
+    certificateVerifier: Opt[CertificateVerifier] = Opt.none(CertificateVerifier),
+): TLSConfig {.gcsafe, raises: [QuicConfigError].} =
+  # In a config, certificate and keys are optional, but if using them, both must
+  # be specified at the same time
+  if certificate.len != 0 or key.len != 0:
+    if certificate.len == 0:
+      raise newException(QuicConfigError, "certificate is required in TLSConfig")
+
+    if key.len == 0:
+      raise newException(QuicConfigError, "key is required in TLSConfig")
+
+  return TLSConfig(
+    certificate: certificate, key: key, certificateVerifier: certificateVerifier
+  )
+
+proc init*(
+    t: typedesc[QuicServer], tlsConfig: TLSConfig
+): QuicServer {.raises: [QuicConfigError].} =
```

Comment:
since `TLSConfig` is created using `init` it will ensure that valid `certificate` and `key` is specified. therfore does it really make sense to rise `QuicConfigError` here?